src/embassy/time_driver.rs

```diff
@@ -0,0 +1,394 @@
+use core::cell::Cell;
+use core::convert::TryInto;
+use core::sync::atomic::{compiler_fence, AtomicU32, AtomicU8, Ordering};
+use core::u16;
+use critical_section::CriticalSection;
+use embassy_sync::blocking_mutex::{raw::CriticalSectionRawMutex, Mutex};
+use embassy_time_driver::{AlarmHandle, Driver, TICK_HZ};
+
+use crate::pac::{interrupt, Interrupt};
+use crate::rcu::{sealed::RcuBus, Clocks, Enable, GetBusFreq, Reset};
+
+#[cfg(not(any(
+    feature = "time-driver-tim1",
+    feature = "time-driver-tim2",
+    feature = "time-driver-tim14",
+)))]
+compile_error!("Embassy feature enabled however a time driver was not specified. A `--features time-driver-tim<1, 2, 14>` is required.");
+
+#[cfg(any(
+    all(feature = "time-driver-tim1", feature = "time-driver-tim2"),
+    all(feature = "time-driver-tim1", feature = "time-driver-tim14"),
+    all(feature = "time-driver-tim2", feature = "time-driver-tim14"),
+))]
+compile_error!(
+    "Multiple Embassy time drivers are specified. Only a single `--features time-driver-tim<1, 2, 14>` can be specified."
+);
+
+#[cfg(feature = "time-driver-tim1")]
+pub(super) type Timer = crate::pac::Timer1;
+#[cfg(feature = "time-driver-tim2")]
+pub(super) type Timer = crate::pac::Timer2;
+#[cfg(feature = "time-driver-tim14")]
+pub(super) type Timer = crate::pac::Timer14;
+
+#[cfg(any(feature = "time-driver-tim1", feature = "time-driver-tim2"))]
+pub(super) type Bus = crate::rcu::APB1;
+#[cfg(feature = "time-driver-tim14")]
+pub(super) type Bus = crate::rcu::APB2;
+
+#[cfg(any(feature = "time-driver-tim1", feature = "time-driver-tim2"))]
+type RegisterBlock = crate::pac::timer1::RegisterBlock;
+#[cfg(feature = "time-driver-tim14")]
+type RegisterBlock = crate::pac::timer14::RegisterBlock;
+
+// NOTE regarding the ALARM_COUNT:
+//
+// This driver is implemented using CC0 as the halfway rollover interrupt, and any additional CC
+// capabilities to provide timer alarms to embassy-time.
+// Hence, only general-purpose timer instances with two or more channels are allowed:
+// - TIMER1, TIMER2 - 4 channels
+// - TIMER14 - 2 channels
+
+#[cfg(any(feature = "time-driver-tim1", feature = "time-driver-tim2"))]
+const ALARM_COUNT: usize = 3;
+
+#[cfg(feature = "time-driver-tim14")]
+const ALARM_COUNT: usize = 1;
+
+fn timer() -> &'static RegisterBlock {
+    unsafe { &*Timer::PTR }
+}
+
+struct AlarmState {
+    timestamp: Cell<u64>,
+    callback: Cell<Option<(fn(*mut ()), *mut ())>>,
+}
+
+unsafe impl Send for AlarmState {}
```

Comment:
If this is sound then please add a safety comment explaining how `AlarmState` is guaranteed to meet the safety requirements of `Send`.